hal/ia32/console-vga.c

```diff
@@ -0,0 +1,294 @@
+								halconsole_common.parms[i] = 0;
+							halconsole_common.parmi = 0;
+							halconsole_common.esc = esc_csi;
+							break;
+
+						default:
+							halconsole_common.esc = esc_init;
+							break;
+					}
+					break;
+
+				case esc_csi:
+					switch (c) {
+						case '0':
+						case '1':
+						case '2':
+						case '3':
+						case '4':
+						case '5':
+						case '6':
+						case '7':
+						case '8':
+						case '9':
+							halconsole_common.parms[halconsole_common.parmi] *= 10;
+							halconsole_common.parms[halconsole_common.parmi] += c - '0';
+							break;
+
+						case ';':
+							halconsole_common.parmi++;
+							break;
+
+						case '?':
+							halconsole_common.esc = esc_csiqm;
+							break;
+
+						case 'H':
+							if (halconsole_common.parms[0] < 1)
+								halconsole_common.parms[0] = 1;
+							else if (halconsole_common.parms[0] > halconsole_common.rows)
+								halconsole_common.parms[0] = halconsole_common.rows;
+
+							if (halconsole_common.parms[1] < 1)
+								halconsole_common.parms[1] = 1;
+							else if (halconsole_common.parms[1] > halconsole_common.cols)
+								halconsole_common.parms[1] = halconsole_common.cols;
+
+							console_setCursor(halconsole_common.parms[0] - 1, halconsole_common.parms[1] - 1);
+							halconsole_common.esc = esc_init;
+							break;
+
+						case 'J':
+							if (halconsole_common.parms[0] < 3) {
+								console_getCursor(&row, &col);
+
+								if (!halconsole_common.parms[0]) {
+									n = (halconsole_common.rows - row - 1) * halconsole_common.cols + halconsole_common.cols - col - 1;
+								}
+								else {
+									console_setCursor(0, 0);
+
+									if (halconsole_common.parms[0] == 1)
+										n = row * halconsole_common.cols + col + 1;
+									else
+										n = halconsole_common.rows * halconsole_common.cols;
+								}
+
+								for (i = 0; i < n; i++)
+									console_write(halconsole_common.attr, ' ');
+
+								console_setCursor(row, col);
+							}
+							halconsole_common.esc = esc_init;
+							break;
+
+						case 'm':
+							i = 0;
+							do {
+								switch (halconsole_common.parms[i]) {
+									case 0:
+										halconsole_common.attr = 0x07;
+										break;
+
+									case 1:
+										halconsole_common.attr = 0x0f;
+										break;
+
+									case 30:
+									case 31:
+									case 32:
+									case 33:
+									case 34:
+									case 35:
+									case 36:
+									case 37:
+										halconsole_common.attr = (halconsole_common.attr & 0xf0) | ansi2fg[(halconsole_common.parms[i] - 30) & 0x7];
+										break;
+
+									case 40:
+									case 41:
+									case 42:
+									case 43:
+									case 44:
+									case 45:
+									case 46:
+									case 47:
+										halconsole_common.attr = ansi2bg[(halconsole_common.parms[i] - 40) & 0x7] | (halconsole_common.attr & 0x0f);
+										break;
+								}
+							} while (i++ < halconsole_common.parmi);
```

Comment:
`parmi` holds number of escape sequence arguments - 1 so it won't overflow as long `parmi` itself is smaller than `sizeof(parms)`. In order to avoid that problem I've added checks for `parmi` incrementation:
```c
if (tty->parmi + 1 < sizeof(tty->parms))
	tty->parmi++;
```